Quantifiers

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
fix typo quanitfiers -> quantifiers

This fixes a typo in the name of the directory of one of the tests.

kroening committed:
don't fail on quantifiers with more than one symbol

This adds support for eager instantiation of forall or exists quantifier
expressions that have more than one symbol using the trivial rewrite.

kroening committed:
add expansion for quantifiers over booleans

This adds support for eager instantiation of quantifiers over booleans,
using George Boole's Proposition II.

kroening committed:
use quantifier_exprt::instantiate instead of replace_expr

The quantifier_exprt offers a convenient wrapper for doing the
instantiation.

kroening committed:
add a test for quantifiers over bit vectors

The test currently fails, and will need an implementation.

tautschnig commented on src/solvers/flattening/boolbv_quantifier.cpp:
```diff
@@ -141,6 +141,16 @@ get_quantifier_var_max(const exprt &var_expr, const exprt &quantifier_expr)
+    new_variables.pop_back();
+    auto new_expression =
+      quantifier_exprt(expr.id(), expr.variables().back(), expr.where());
+    return eager_quantifier_instantiation(new_expression, ns);
```
This function doesn't exist at the point of this commit (else line 142 should already say "eager_quantifier_instantiation").

kroening replied:
This will go way once rebased on #6551.

kroening replied:
Now fixed with separate commit.

tautschnig commented on src/solvers/flattening/boolbv_quantifier.cpp:
```diff
@@ -141,6 +141,16 @@ get_quantifier_var_max(const exprt &var_expr, const exprt &quantifier_expr)
+    auto new_variables = expr.variables();
+    new_variables.pop_back();
+    auto new_expression =
+      quantifier_exprt(expr.id(), expr.variables().back(), expr.where());
```
So what about all the other variables, don't they just get dropped? Shouldn't `expr.where()` be replaced by `quantifier_exprt{expr.id(), new_variables, expr.where()}`?

kroening replied:
yes, fixed

tautschnig commented on src/solvers/flattening/boolbv_quantifier.cpp:
```diff
@@ -169,58 +169,80 @@ static optionalt<exprt> eager_quantifier_instantiation(
+    else
+      UNREACHABLE;
+  }
+  else
+  {
```
We could reduce code churn by removing the `else` here?

kroening replied:
ok

tautschnig commented on regression/smt2_solver/quantifiers/bv-quantifier.smt2:
```diff
@@ -0,0 +1,16 @@
+(define-fun q1 () Bool (exists ((b (_ BitVec 8))) (= b #x00)))
+;(define-fun q2 () Bool (exists ((b (_ BitVec 8))) (not (= b #x00))))
+
+;(define-fun q3 () Bool (not (forall ((b (_ BitVec 8))) (= b #x00))))
+;(define-fun q4 () Bool (not (forall ((b (_ BitVec 8))) (not (= b #x00)))))
+
+;(define-fun q5 () Bool (exists ((a (_ BitVec 8)) (b (_ BitVec 8))) (= a b)))
+;(define-fun q6 () Bool (not (forall ((a (_ BitVec 8)) (b (_ BitVec 8))) (= a b))))
+
+;(define-fun q7 () Bool (forall ((a (_ BitVec 8))) (exists ((b (_ BitVec 8))) (= a b))))
+
+; the above are all valid, and we assert one of them is not
+;(assert (not (and q1 q2 q3 q4 q5 q6 q7)))
```
Why are all of these commented out?

kroening replied:
Now marked as a KNOWNBUG test.

tautschnig reviewed (changes requested):
I believe there are multiple issues in the commit introducing support for multiple variables as detailed below.

kroening committed:
rename instantiate_quantifier function

Quantifier instantiation may happen eagerly, or incrementally, and renaming
this function clarifies that this function is doing eager instantiation.

tautschnig commented on regression/smt2_solver/quantifiers/bv-quantifier-valid.desc:
```diff
@@ -0,0 +1,7 @@
+KNOWNBUG
+bv-quantifier-valid.smt2
+
+^EXIT=0$
+^SIGNAL=0$
+^unsat$
+--
```
It would be helpful for either this file or the commit message to include an explanation what exactly is missing here.
